Refactored and made the code look better

RomaDevWorld opened the pull request:
So I basically removed highly deeply nested code
I also reformatted the code using prettier to make it more readable, and ESLint to fix some errors

**I recommend you cloning my code and testing it before merging.**

Have a good one!

RomaDevWorld committed:
Add: Prettier config

RomaDevWorld committed:
Add: .gitignore

RomaDevWorld committed:
Add: ESLint config & lint script

RomaDevWorld committed:
Update: prettier config

RomaDevWorld committed:
Update: README.md

RomaDevWorld committed:
Update: Several files

Removed deeply nesting, added some return conditions instead of if/else

Arisamiga commented on events/messageCreate.js:
```diff
@@ -1,28 +1,25 @@
-  
+
+    // if (message.content.indexOf(prefix) !== 0) return;
+    if (message.content.startsWith(prefix)) return;
```
This is supposed to return when the content doesn't have the prefix

RomaDevWorld replied:
Fixed it.

Arisamiga commented on main.js:
```diff
@@ -94,39 +79,44 @@ client.on('ready', async () => {
 client.login(token.token);
+
+const checkMessage = async (id, channelid) => {
+        const channel = await client.channels.cache.fetch(channelid);
```
We don't want to use cache here for two reasons:
+ 1st client.channels.cache.fetch is not a function
+ Is so we can get messages before the bot has been online

RomaDevWorld replied:
Fixed

Arisamiga commented on main.js:
```diff
@@ -94,39 +79,44 @@ client.on('ready', async () => {
+
+const checkMessage = async (id, channelid) => {
+        const channel = await client.channels.cache.fetch(channelid);
+        const channelMessages = await channel.messages.fetch()
```
We are also missing input into the fetch function because we are searching for 1 specific id

RomaDevWorld replied:
fixed

Arisamiga commented:
Thank you for contributing to the repository! :D

There seem to be some problems that are preventing the bot from running accordingly.

I have outlined some of the problems above but other problems are:

+ Bot always thinks there is an ongoing ticket and because of that, it doesn't create a channel.
+ Bot deleting collector even if it exists
+ Reactions not being handled correctly
+ eslint not being configured with prettier so it could conflict with prettier
+ Some of the if statements are incorrectly handled

If these changes worked on your setup please be sure that you are including all the changes 🙂

Arisamiga reviewed (changes requested):
Thank you for contributing to the repository! :D

There seem to be some problems that are preventing the bot from running accordingly.

I have outlined some of the problems above but other problems are:

+ Bot always thinks there is an ongoing ticket and because of that, it doesn't create a channel.
+ Bot deleting collector even if it exists
+ Reactions not being handled correctly
+ eslint not being configured with prettier so it could conflict with prettier
+ Some of the if statements are incorrectly handled

If these changes worked on your setup please be sure that you are including all the changes 🙂

RomaDevWorld committed:
Fixed "checkMessage" function in main.js

RomaDevWorld committed:
Fixed messageCreate prefix issue

RomaDevWorld committed:
Update: eslint config

RomaDevWorld committed:
Fix: messageReactionAdd

RomaDevWorld commented:
I made some changes

1. Fixed `checkMessage` in `main.js` by adding `id` property to `messages.fetch` and removing `cache` between `channels.fetch`
2. Fixed `message.content.startsWith` in `messageCreate.ts` by adding a singe exclamation mark
3.  Added prettier plugin to `.eslintrc.js` so it will work properly with prettier

Arisamiga commented on events/messageReactionAdd.js:
```diff
@@ -1,87 +1,71 @@
+    if (messageReaction.emoji.name === '📩') {
+      let channelname = `ticket-${user.username}`;
+      channelname = channelname.replace(/\s/g, '-').toLowerCase();
+      if (!messageReaction.message.guild.channels.cache.find((channel) => channel.name === channelname) || !config.one_app) {
```
This if statement should not have !'s and should not be a ||

Arisamiga reviewed (changes requested):
There are still some issues that I have highlighted above but also the following:
+ There are a lot of if statements that have "!" that shouldn't be there
+ Bot deleting collector even if it exists
+ Some of the if statements are incorrectly handled
+ Reactions are not being handled based on the role

Please be sure to first test it before requesting a review 🙂

RomaDevWorld committed:
Fix: checkMessage now returns message object instead of true or false

RomaDevWorld commented:
So.. Tell me again, why do you prefer using `client.channels.fetch` instead of `client.channels.cache.get`?

The thing is that in the first case if the channel doesn't exist it will throw an error and crash an entire application.
I'll use `client.channels.cache.get` since it works just fine :)

RomaDevWorld committed:
Changed channels.fetch to channels.cache.get

RomaDevWorld committed:
Fixed incorrect if statement

RomaDevWorld commented:
1. Collectors and tickets won't be deleted if they still exist 
2. Only authorized roles and “Admins” can close/delete ticket
3. The misunderstood "if" statement was fixed and now works as expected.

P.S: I've tested it, and haven't found any issues

Fell free to point me to any new problems you'll find!

Arisamiga commented on main.js:
```diff
@@ -1,132 +1,121 @@
+const checkMessage = async (id, channelid) => {
+  const channel = await client.channels.cache.get(channelid);
+  if (!channel) return;
+  const channelMessage = await channel?.messages.fetch(id);
```
Might want to add a .catch to prevent the application from crashing because if there is an error it is most likely because the message doesn't exist.

RomaDevWorld replied:
if message doesn't exist the `channelMessage` will just be undefined. No error here.

Arisamiga replied:
If the message doesn't exist the channelMessage will crash the application because Discord returns a 404 error code with `DiscordAPIError: Unknown Message`  so I suggest adding a `.catch` there.

Arisamiga commented on main.js:
```diff
@@ -1,132 +1,121 @@
 client.login(token.token);
+
+const checkMessage = async (id, channelid) => {
+  const channel = await client.channels.cache.get(channelid);
```
Might want to add a .catch to prevent the application from crashing because if there is an error it is most likely because the channel doesn't exist so we want to remove the messages from that channel

RomaDevWorld replied:
if the channel doesn't exist on property `client.channels.cache` then `channel` will just be undefined. 
On the next line `if (!channel) return;` will stop the function from executing further. No error here.

Arisamiga commented on events/messageReactionAdd.js:
```diff
@@ -1,87 +1,71 @@
+    let channel = messageReaction.message.guild.channels.cache.find((channel) => channel.id === messageReaction.message.channel.id);
+    switch (messageReaction.emoji.name) {
+      case '🔒':
+        if (!checkUser(messageReaction.message, user) || !config.allow_user_lock) {
```
Since this uses "!" you should use "&&" not "||"

RomaDevWorld replied:
Why tho? Am I missing something? 

If the user is not authorized to close the channel, or config doesn't allow this then respond with "Only Staff can lock the channels"

RomaDevWorld replied:
Oh, I see.

> (Allow users to lock their **own** tickets)

Arisamiga commented on events/messageReactionAdd.js:
```diff
@@ -1,87 +1,71 @@
+        return messageReaction.message.channel.send('Channel Locked 🔒');
+      case '⛔':
+        if (!messageReaction.message.guild.channels.cache.find((c) => c.name.toLowerCase() === channel.name)) return;
+        if (!checkUser(messageReaction.message, user) || !config.allow_user_delete) {
```
Since this uses "!" you should use "&&" not "||"

Arisamiga reviewed (changes requested):
There are some minor issues which I have outlined above!

Also might wanna install the config for prettier: `ESLint couldn't find the config "prettier" to extend from.` :)

RomaDevWorld committed:
Update: messageReactionAdd.js

RomaDevWorld committed:
Update: Package list and scripts

RomaDevWorld committed:
Update: package.json

RomaDevWorld committed:
Add: catch block

RomaDevWorld committed:
Update: nodemon config

RomaDevWorld commented:
Made the changes that you mentioned.
Also added nodemon to automatically refresh the application on file save
`npm run dev`

Arisamiga reviewed (approved):
Thanks so much for contributing 🙂 Everything is working Great!

RomaDevWorld commented:
I also recomend you learning and using typescript in your future javascript projects. It can be a little bit hard to learn and understand, but when you do it will make your projects 10x better!

Stay safe!
